numpy/ma/core.pyi

```diff
@@ -347,13 +349,13 @@ class MaskedArray(ndarray[_ShapeType_co, _DType_co]):
     def __getitem__(self, indx): ...
     def __setitem__(self, indx, value): ...
     @property
-    def dtype(self): ...
+    def dtype(self) -> _DType_co: ...
     @dtype.setter
-    def dtype(self, dtype): ...
+    def dtype(self: MaskedArray[_ShapeType_co, _DType], dtype: _DType) -> None: ...
     @property
-    def shape(self): ...
+    def shape(self) -> _ShapeType_co: ...
     @shape.setter
-    def shape(self, shape): ...
+    def shape(self: MaskedArray[_ShapeType, _DType_co], shape: _ShapeType) -> None: ...
```

Comment:
Haha, I came up with this same trick to get out the covariance in numtype: https://github.com/numpy/numtype/blob/131a491389393842f91d7b1ae236aa42a6067827/src/numpy-stubs/ma/core.pyi#L309-L320

Quick nitpick: it's slightly cleaner to make the setter parameters positional-only

Reply:
😄 I had a look at that first for reference

tbh it looks very odd to me that NumPy allows `@shape.setter`, but that's a separate discussion

Reply:
> tbh it looks very odd to me that NumPy allows `@shape.setter`, but that's a separate discussion

I agree